src/main/java/com/google/cloud/teleport/splunk/SplunkEventWriter.java

```diff
@@ -386,9 +510,82 @@ public Builder withToken(ValueProvider<String> token) {
      * @return {@link Builder}
      */
     public Builder withToken(String token) {
-      checkArgument(token != null, "withToken(token) called with null input.");
       return setToken(ValueProvider.StaticValueProvider.of(token));
     }
+  
+    /**
+     * Method to set the KMS Encryption Key for HEC token.
+     *
+     * @param tokenKmsEncryptionKey KMS Encryption Key for HEC token
+     * @return {@link Builder}
+     */
+    public Builder withTokenKmsEncryptionKey(ValueProvider<String> tokenKmsEncryptionKey) {
+      return setTokenKmsEncryptionKey(tokenKmsEncryptionKey);
+    }
+  
+    /**
+     * Same as {@link Builder#withTokenKmsEncryptionKey(ValueProvider)} but without {@link ValueProvider}.
+     *
+     * @param tokenKmsEncryptionKey KMS Encryption Key for HEC token
+     * @return {@link Builder}
+     */
+    public Builder withTokenKmsEncryptionKey(String tokenKmsEncryptionKey) {
+      return setTokenKmsEncryptionKey(ValueProvider.StaticValueProvider.of(tokenKmsEncryptionKey));
+    }
+  
+    /**
+     * Method to set the Secret Manager Secret Id for HEC token.
+     *
+     * @param tokenSecretId Secret Manager Secret Id for HEC token
+     * @return {@link Builder}
+     */
+    public Builder withTokenSecretId(ValueProvider<String> tokenSecretId) {
+      return setTokenSecretId(tokenSecretId);
+    }
+  
+    /**
+     * Same as {@link Builder#withTokenSecretId(ValueProvider)} but without {@link ValueProvider}.
+     *
+     * @param tokenSecretId Secret Manager Secret Id for HEC token
+     * @return {@link Builder}
+     */
+    public Builder withTokenSecretId(String tokenSecretId) {
+      return setTokenSecretId(ValueProvider.StaticValueProvider.of(tokenSecretId));
+    }
+  
+    /**
+     * Source of the token. One of PLAINTEXT, KMS or SECRET_MANAGER.
+     *
+     * @param tokenSource Source of the token
+     * @return {@link Builder}
+     */
+    public Builder withTokenSource(ValueProvider<String> tokenSource) {
+      checkArgument(tokenSource != null, "withTokenSource(tokenSource) called with null input");
+  
+      if (tokenSource.isAccessible() && !EnumUtils.isValidEnum(SplunkTokenSource.class, tokenSource.get())) {
```

Comment:
`with*` methods and some validation logic (e.g. checking for `isValidEnum`) are still duplicated in `SplunkIO` and `SplunkEventWriter`.
I'm OK with submitting this as is, but my recommendation would be to move `resolveTokenSource` and `getToken` to `SplunkIO`, resolve the token there, and pass an already resolved token to `SplunkEventWriter`. Then you won't need to duplicate `withTokenSource`, `withTokenSecretId`, `withTokenKmsEncryptionKey` and their validation logic here. This is also better from the "fail fast" perspective.

Reply:
We may not be able to build a template for the pipeline if try to resolve it in `SplunkIO`. The code inside `SplunkIO`'s `expand()` method is evaluated while building the template and since the token params won't be present at build time, it'll fail with an exception.

Maybe we should go ahead with a new `TokenNestedValueProvider` class that accepts the token params and outputs the final resolved token like you suggested offline? Might need a little more refactoring from the current approach but I agree it would be cleaner.